v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/failureinjectiontesting/SourceDbToSpannerFTBase.java

```diff
@@ -0,0 +1,279 @@
+        JSONObject shardJson = new JSONObject();
+
+        shardJson.put("dataShardId", shardData.dataShardId);
+        shardJson.put("host", shardData.host);
+        shardJson.put("user", shardData.user);
+        shardJson.put("password", shardData.password);
+        shardJson.put("port", shardData.port);
+        shardJson.put("dbName", shardData.dbName);
+        shardJson.put("namespace", shardData.namespace);
+        shardJson.put("connectionProperties", shardData.connectionProperties);
+
+        JSONArray databasesArray = new JSONArray();
+
+        for (Database dbData : shardData.databases) {
+          JSONObject dbJson = new JSONObject();
+          dbJson.put("dbName", dbData.dbName);
+          dbJson.put("databaseId", dbData.databaseId);
+          dbJson.put("refDataShardId", dbData.refDataShardId);
+          databasesArray.put(dbJson);
+        }
+        shardJson.put("databases", databasesArray);
+        dataShardsArray.put(shardJson);
+      }
+    }
+    shardConfigBulk.put("dataShards", dataShardsArray);
+
+    bulkConfig.put("shardConfigurationBulk", shardConfigBulk);
+    String shardFileContents = bulkConfig.toString();
+    LOG.info("Shard file contents: {}", shardFileContents);
+    gcsResourceManager.createArtifact("input/shard-bulk.json", shardFileContents);
+  }
+
+  protected class DataShard {
+    String dataShardId;
+    String host;
+    String user;
+    String password;
+    String port;
+    String dbName;
+    String namespace;
+    String connectionProperties;
+    ArrayList<Database> databases;
+
+    public DataShard(
+        String dataShardId,
+        String host,
+        String user,
+        String password,
+        String port,
+        String dbName,
+        String namespace,
+        String connectionProperties,
+        ArrayList<Database> databases) {
+
+      this.dataShardId = dataShardId;
+      this.host = host;
+      this.user = user;
+      this.password = password;
+      this.port = port;
+      this.dbName = dbName;
+      this.namespace = namespace;
+      this.connectionProperties = connectionProperties;
+      this.databases = databases;
+    }
+  }
+
+  protected class Database {
+    String dbName;
+    String databaseId;
+    String refDataShardId;
+
+    public Database(String dbName, String databaseId, String refDataShardId) {
+      this.dbName = dbName;
+      this.databaseId = databaseId;
+      this.refDataShardId = refDataShardId;
+    }
+  }
+
+  public static class SevereErrorsCheck extends ConditionCheck {
+
+    private Integer minErrors;
+
+    private PipelineLauncher.LaunchInfo jobInfo;
+
+    private PipelineLauncher pipelineLauncher;
+
+    @Override
+    public String getDescription() {
+      return String.format("Severe errors check if %d errors are present", minErrors);
+    }
+
+    @Override
+    public CheckResult check() {
+      Double severeErrors = null;
+      try {
+        severeErrors =
+            pipelineLauncher.getMetric(
```

Comment:
Wondering whether we rely on the metric, or should we rely on the number of entries in GCS?

If we want to have a single check,
I might lean on GCS - as if we have a metric and not file in GCS there's data loss.
Also as per the general best practice of testing behavior that's critical testing GCS is better than the metric.